qep-323-code-submission-policy.md

```diff
@@ -0,0 +1,53 @@
+# QGIS Enhancement 323: Change submission and merge policies
+
+## Summary
+
+This QEP documents policies for submission of changes to the main QGIS code repository.
+Developers are required to follow these standards when submitting code to QGIS.
+
+## Policies
+
+### 1. Change submission
+
+- 1.1. All changes must be submitted in the form of Pull Requests via GitHub.
+  - 1.1.1. Pull Requests can be made by **ANYONE**, and **EVERYONE** is encouraged to
+    contribute to QGIS coding! No special permission or status is required to submit
+    a change to QGIS.
+  - 1.1.2. Direct push to master is permitted in **extreme circumstances** only. These are:
+    - String fixes for code comments or other non-functional text. (This does not include
+      code documentation, changes to the README files, or other documentation).
+    - Fixes for the CI environment or GitHub workflows where the lengthy run-time of
+      CI checks on pull requests is prohibitive, or where there is a need to repair
+      CI environment in order to allow normal pull request activity.
+    - Release and packaging related commits, such as version tagging.
+  - 1.1.3. All Pull Requests should be accompanied by a descriptive title and explanatory text.
+    Linking to existing issues, feature requests or external discussion are encouraged
+    for reviewer context. 
+- 1.2. All code submissions must be accompanied by appropriate unit tests.
+  - 1.2.1. All changes to the ``core``, ``analysis``, or ``server`` modules must
+    have extensive unit testing.
+    - 1.2.1.1. Exceptions are made where the effort required in mocking interfaces is excessive,
+      or where tests are likely to be flaky and trigger false-positives on the CI environment.
+  - 1.2.2. Changes to the ``gui`` or ``app`` modules should have as many tests as are possible
+    without extensive GUI interface mocking.
+    - 1.2.2.1. There are no expectations for testing of keyboard or mouse interaction with
+      dialogs and widgets.
+  - 1.2.3. Changes to data providers must be accompanied by appropriate unit tests, unless
+    the effort required to mock external interfaces is excessive.
+- 1.3. Code changes must abide by the [coding style policies](qep-314-coding-style.md).
+
+### 2. Pull Request Review and Merging
+
+Before a Pull Request can be merged to the repository, it must:
+
+- 2.1. Pass all existing CI checks.
+- 2.2. Discussions and comments from any interested party are encouraged. No special
+  status or permissions are required to participate in the review discussion.
+- 2.3. Before merge, the Pull Request **MUST** be reviewed and approved by a **Core Contributor**
+  with repository write permission.
+  - 2.3.1. Developers are **NOT** permitted to self-approve Pull Requests they have submitted.
+    - 2.3.1.1. An exception is made where a developer has submitted a "fixed up" version
```

Comment:
May I add an exception for when you are cleaning your own mess and you are 100% sure of what you are doing and nobody else approved or even commented your PR after one week ?

These are exceptional circumstances and shouldn't happen in a perfect world but it happened to me a few times.

Reply:
@elpaso given that this is reflection of CURRENT practices and that this does happen, then yes, let's add it. Got any ideas on how we can write that in a semi formal way?

Reply:
Difficult task...

The current situation IMO is that experienced QGIS developers (let's say > 5 years on the QGIS job) might exceptionally self approve a PR in case they are fixing a bug in code written by themselves or when the changes are minor/trivial and where they are absolutely sure there isn't any danger of breaking things, **if and only if their PR didn't get any comment/approval in about one week**.

We can argue if this is a good or a bad practice but it happened to me and in full conscience I believe I did it for the good of the project and since we are trying to capture the current practice, here it is.

Reply:
Ok, how's [af92f98](https://github.com/qgis/QGIS-Enhancement-Proposals/pull/329/commits/af92f984492371583f7d8f1e0d35a2673e093970) look?

Reply:
Looks good, thanks!